# Lab 07: Data Cleaning & Data Transformation

In this lab, you will:

- Import "raw" data and clean the imported data
- Transform the data

**Deadline: 11:59 PM Tuesday 10/26/2021**

***

## 1. Importing Data

In this lab, we will investigate the U.S. Small Business Administration (SBA) loan dataset. The dataset details can be found on [Kaggle](https://www.kaggle.com/mirbektoktogaraev/should-this-loan-be-approved-or-denied).

- Make sure you take a close look at the list of variables!
- You will need to download the dataset to your local computer, then use the local file to load the data to Python.
- **DO NOT STORE THE DATA ANYWHERE INSIDE YOUR NETID FOLDER!!!**
    - This dataset is too big for GitHub and you won't be able to push to the remote repo.

Import the data to a DataFrame named `sba_loan`.

In [1]:
import pandas as pd
sba_loan = pd.read_csv("C:/Users/92963/Desktop/1/archive/SBAnational.csv")
sba_loan

C:\Users\92963\miniconda3\lib\site-packages\IPython\core\interactiveshell.py:3441: DtypeWarning: Columns (9) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


LoanNr_ChkDgt                            Name             City State  \
0          1000014003                  ABC HOBBYCRAFT       EVANSVILLE    IN   
1          1000024006     LANDMARK BAR & GRILLE (THE)        NEW PARIS    IN   
2          1000034009           WHITLOCK DDS, TODD M.      BLOOMINGTON    IN   
3          1000044001   BIG BUCKS PAWN & JEWELRY, LLC     BROKEN ARROW    OK   
4          1000054004     ANASTASIA CONFECTIONS, INC.          ORLANDO    FL   
...               ...                             ...              ...   ...   
899159     9995573004                    FABRIC FARMS  UPPER ARLINGTON    OH   
899160     9995603000                    FABRIC FARMS         COLUMBUS    OH   
899161     9995613003    RADCO MANUFACTURING CO.,INC.      SANTA MARIA    CA   
899162     9995973006           MARUTAMA HAWAII, INC.         HONOLULU    HI   
899163     9996003010  PACIFIC TRADEWINDS FAN & LIGHT           KAILUA    HI   

          Zip                            Bank BankState   NAICS ApprovalDate  \
0       47711                FIFTH THIRD BANK        OH  451120    28-Feb-97   
1       46526                 1ST SOURCE BANK        IN  722410    28-Feb-97   
2       47401         GRANT COUNTY STATE BANK        IN  621210    28-Feb-97   
3       74012   1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97   
4       32801         FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97   
...       ...                             ...       ...     ...          ...   
899159  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451120    27-Feb-97   
899160  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451130    27-Feb-97   
899161  93455  RABOBANK, NATIONAL ASSOCIATION        CA  332321    27-Feb-97   
899162  96830                  BANK OF HAWAII        HI       0    27-Feb-97   
899163  96734            CENTRAL PACIFIC BANK        HI       0    27-Feb-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
0            1997  ...          N       Y         NaN         28-Feb-99   
1            1997  ...          N       Y         NaN         31-May-97   
2            1997  ...          N       N         NaN         31-Dec-97   
3            1997  ...          N       Y         NaN         30-Jun-97   
4            1997  ...          N       N         NaN         14-May-97   
...           ...  ...        ...     ...         ...               ...   
899159       1997  ...          0       N         NaN         30-Sep-97   
899160       1997  ...          Y       N         NaN         31-Oct-97   
899161       1997  ...          N       N         NaN         30-Sep-97   
899162       1997  ...          N       Y    8-Mar-00         31-Mar-97   
899163       1997  ...          N       N         NaN         31-May-97   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr  \
0             $60,000.00         $0.00        P I F       $0.00    
1             $40,000.00         $0.00        P I F       $0.00    
2            $287,000.00         $0.00        P I F       $0.00    
3             $35,000.00         $0.00        P I F       $0.00    
4            $229,000.00         $0.00        P I F       $0.00    
...                   ...           ...         ...          ...   
899159        $70,000.00         $0.00        P I F       $0.00    
899160        $85,000.00         $0.00        P I F       $0.00    
899161       $300,000.00         $0.00        P I F       $0.00    
899162        $75,000.00         $0.00       CHGOFF  $46,383.00    
899163        $30,000.00         $0.00        P I F       $0.00    

              GrAppv      SBA_Appv  
0        $60,000.00    $48,000.00   
1        $40,000.00    $32,000.00   
2       $287,000.00   $215,250.00   
3        $35,000.00    $28,000.00   
4       $229,000.00   $229,000.00   
...              ...           ...  
899159   $70,000.00    $56,000.00   
899160   $85,000.00    $42,500.00   
899161  $300,000.00   $225,000.00  

### Exercise 1.1

Take a look at the data types of each column and compare it with the corresponding variable description. For each column, decide if you need to modify the data type or keep it as it is. Include your reasoning.

*Note: There are many possible correct answers here. Your reasoning is what matters.*

#### Your Answer:

*Please use bullet point!*
 
*  **column17  RevLineCr**         
    reason: modify, the result should be Boolean value
*  **column18  LowDoc**           
    reason: modify, the result should be Boolean value
  
*  **column21  DisbursementGross**    
    reason: modify,remove the dollar sign and the result should be float
*  **column22  BalanceGross**       
    reason: modify,remove the dollar sign and the result should be float
*  **column24  ChgOffPrinGr**       
    reason: modify,remove the dollar sign and the result should be float
*  **column25  GrAppv**           
    reason: modify,remove the dollar sign and the result should be float
*  **column26  SBA_Appv**          
    reason: modify,remove the dollar sign and the result should be float

### Exercise 1.2

Based on your answer on 1.1, perform any needed modification.

In [2]:
# column
def toBoolean(a):
    if (a == 'Y'):
        return 1
    elif(a == 'N'):
        return 0

sba_loan['RevLineCr'] = sba_loan.apply(lambda x:toBoolean(x.RevLineCr),axis = 1)
sba_loan['LowDoc'] = sba_loan.apply(lambda x:toBoolean(x.LowDoc),axis = 1)

sba_loan['DisbursementGross']  = sba_loan['DisbursementGross'].replace({'\$':'',",":""}, regex = True).astype(float)
sba_loan['BalanceGross']  = sba_loan['BalanceGross'].replace({'\$':'',",":""}, regex = True).astype(float)
sba_loan['ChgOffPrinGr']  = sba_loan['ChgOffPrinGr'].replace({'\$':'',",":""}, regex = True).astype(float)
sba_loan['GrAppv']  = sba_loan['GrAppv'].replace({'\$':'',",":""}, regex = True).astype(float)
sba_loan['SBA_Appv']  = sba_loan['SBA_Appv'].replace({'\$':'',",":""}, regex = True).astype(float)
sba_loan[['RevLineCr','LowDoc','DisbursementGross','BalanceGross','ChgOffPrinGr','GrAppv','SBA_Appv']]

NameError: name 'sba_loan' is not defined

### Exercise 1.3

Identify the columns with missing data.

In [3]:
# your code
sba_loan.isnull().any(axis = 0)

LoanNr_ChkDgt        False
Name                  True
City                  True
State                 True
Zip                  False
Bank                  True
BankState             True
NAICS                False
ApprovalDate         False
ApprovalFY           False
Term                 False
NoEmp                False
NewExist              True
CreateJob            False
RetainedJob          False
FranchiseCode        False
UrbanRural           False
RevLineCr             True
LowDoc                True
ChgOffDate            True
DisbursementDate      True
DisbursementGross    False
BalanceGross         False
MIS_Status            True
ChgOffPrinGr         False
GrAppv               False
SBA_Appv             False
dtype: bool

***

## 2. Data Transformation

### Exercise 2.1

If you have not removed the dollar sign and change data type of `int64` or `float64` of the following columns in Exercise 1.2, then please do so here!

- `DisbursementGross`
- `BalanceGross`
- `ChgOffPrinGr`
- `SBA_Appv`

In [6]:
# your code here
sba_loan[['DisbursementGross','BalanceGross','ChgOffPrinGr','SBA_Appv']]

DisbursementGross  BalanceGross  ChgOffPrinGr  SBA_Appv
0                 60000.0           0.0           0.0   48000.0
1                 40000.0           0.0           0.0   32000.0
2                287000.0           0.0           0.0  215250.0
3                 35000.0           0.0           0.0   28000.0
4                229000.0           0.0           0.0  229000.0
...                   ...           ...           ...       ...
899159            70000.0           0.0           0.0   56000.0
899160            85000.0           0.0           0.0   42500.0
899161           300000.0           0.0           0.0  225000.0
899162            75000.0           0.0       46383.0   60000.0
899163            30000.0           0.0           0.0   24000.0

[899164 rows x 4 columns]

### Exercise 2.2

Create a new column named `Industry`. Use the first 2 digits of NAICS (if provided) to identify the industry for each observation.

For example, the first observation's NAICS is 451120. The first two digits are 45, so the value for this observation in the `Industry` column is `Retail trade` (see the table provided on Kaggle).

**Notes: The use of loops is not allowed in this exercise. Vectorized string functions such as `str.func_name()` are not allowed either (since it has not been covered in this course, we will learn it in future weeks).**

In [7]:
# your code here

def industry(a):
    if(a == 11):
        return 'Agriculture, forestry, fishing and hunting'
    elif(a == 21):
        return 'Mining, quarrying, and oil and gas extraction'
    elif(a == 22):
        return 'Utilities'
    elif(a == 23):
        return 'Construction'
    elif((a >= 31) & (a <= 33)):
        return 'Manufacturing'
    elif(a == 42):
        return 'Wholesale trade'
    elif((a >= 44) & (a <= 45)):
        return 'Retail trade'
    elif((a >= 48) & (a <= 49)):
        return 'Transportation and warehousing'
    elif(a == 51):
        return 'Information'
    elif(a == 52):
        return 'Finance and insurance'
    elif(a == 53):
        return 'Real estate and rental and leasing'
    elif(a == 54):
        return 'Professional, scientific, and technical services'
    elif(a == 55):
        return 'Management of companies and enterprises'
    elif(a == 56):
        return 'Administrative and support and waste management and remediation services'
    elif(a == 61):
        return 'Educational services'
    elif(a == 62):
        return 'Health care and social assistance'
    elif(a == 71):
        return 'Arts, entertainment, and recreation'
    elif(a == 72):
        return 'Accommodation and food services'
    elif(a == 81):
        return 'Other services (except public administration) 92 Public administration'
    else:
        return 'None'

sba_loan['NAICS'] = sba_loan['NAICS'].astype(str)
sba_loan['NAICS_2'] = (sba_loan['NAICS'].apply(lambda x:x[:2])).astype(int)
sba_loan['Industry'] = sba_loan.apply(lambda x: industry(x.NAICS_2), axis = 1)
sba_loan

LoanNr_ChkDgt                            Name             City State  \
0          1000014003                  ABC HOBBYCRAFT       EVANSVILLE    IN   
1          1000024006     LANDMARK BAR & GRILLE (THE)        NEW PARIS    IN   
2          1000034009           WHITLOCK DDS, TODD M.      BLOOMINGTON    IN   
3          1000044001   BIG BUCKS PAWN & JEWELRY, LLC     BROKEN ARROW    OK   
4          1000054004     ANASTASIA CONFECTIONS, INC.          ORLANDO    FL   
...               ...                             ...              ...   ...   
899159     9995573004                    FABRIC FARMS  UPPER ARLINGTON    OH   
899160     9995603000                    FABRIC FARMS         COLUMBUS    OH   
899161     9995613003    RADCO MANUFACTURING CO.,INC.      SANTA MARIA    CA   
899162     9995973006           MARUTAMA HAWAII, INC.         HONOLULU    HI   
899163     9996003010  PACIFIC TRADEWINDS FAN & LIGHT           KAILUA    HI   

          Zip                            Bank BankState   NAICS ApprovalDate  \
0       47711                FIFTH THIRD BANK        OH  451120    28-Feb-97   
1       46526                 1ST SOURCE BANK        IN  722410    28-Feb-97   
2       47401         GRANT COUNTY STATE BANK        IN  621210    28-Feb-97   
3       74012   1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97   
4       32801         FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97   
...       ...                             ...       ...     ...          ...   
899159  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451120    27-Feb-97   
899160  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451130    27-Feb-97   
899161  93455  RABOBANK, NATIONAL ASSOCIATION        CA  332321    27-Feb-97   
899162  96830                  BANK OF HAWAII        HI       0    27-Feb-97   
899163  96734            CENTRAL PACIFIC BANK        HI       0    27-Feb-97   

       ApprovalFY  ...  ChgOffDate  DisbursementDate  DisbursementGross  \
0            1997  ...         NaN         28-Feb-99            60000.0   
1            1997  ...         NaN         31-May-97            40000.0   
2            1997  ...         NaN         31-Dec-97           287000.0   
3            1997  ...         NaN         30-Jun-97            35000.0   
4            1997  ...         NaN         14-May-97           229000.0   
...           ...  ...         ...               ...                ...   
899159       1997  ...         NaN         30-Sep-97            70000.0   
899160       1997  ...         NaN         31-Oct-97            85000.0   
899161       1997  ...         NaN         30-Sep-97           300000.0   
899162       1997  ...    8-Mar-00         31-Mar-97            75000.0   
899163       1997  ...         NaN         31-May-97            30000.0   

        BalanceGross  MIS_Status  ChgOffPrinGr    GrAppv  SBA_Appv  NAICS_2  \
0                0.0       P I F           0.0   60000.0   48000.0       45   
1                0.0       P I F           0.0   40000.0   32000.0       72   
2                0.0       P I F           0.0  287000.0  215250.0       62   
3                0.0       P I F           0.0   35000.0   28000.0        0   
4                0.0       P I F           0.0  229000.0  229000.0        0   
...              ...         ...           ...       ...       ...      ...   
899159           0.0       P I F           0.0   70000.0   56000.0       45   
899160           0.0       P I F           0.0   85000.0   42500.0       45   
899161           0.0       P I F           0.0  300000.0  225000.0       33   
899162           0.0      CHGOFF       46383.0   75000.0   60000.0        0   
899163           0.0       P I F           0.0   30000.0   24000.0        0   

                                 Industry  
0                            Retail trade  
1         Accommodation and food services  
2       Health care and social assistance  
3                                    None  
4               

In [1]:
industry_dict = {
    11: 'Agriculture, forestry, fishing and hunting',
    21: 'Mining, quarrying, and oil and gas extraction',
    22: 'Utilities',
    23: 'Construction',
    31: 'Manufacturing',
    32: 'Manufacturing',
    33: 'Manufacturing',
    42: 'Wholesale trade',
    44: 'Retail trade',
    45: 'Retail trade',
    48: 'Transportation and warehousing',
    49: 'Transportation and warehousing',
    51: 'Information',
    52: 'Finance and insurance',
    53: 'Real estate and rental and leasing',
    54: 'Professional, scientific, and technical services',
    55: 'Management of companies and enterprises',
    56: 'Administrative and support and waste management and remediation services',
    61: 'Educational services',
    62: 'Health care and social assistance',
    71: 'Arts, entertainment, and recreation',
    72: 'Accommodation and food services',
    81: 'Other services (except public administration)',
    92: 'Public administration'
}

def industry(x):
    try:
        # or x // 10000
        return industry_dict[int(str(x)[:2])]
    except:
        return np.nan

sba_loan['Industry'] = sba_loan['NAICS'].map(industry)
sba_loan

NameError: name 'sba_loan' is not defined

### Exercise 2.3 (bonus)

If you have not changed the data type of the `ApprovalDate` and the `ChgOffDate` columns to `datetime64` in Exercise 1.2, then please do so! The information provided [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html) might be helpful!

In [10]:
# your code here
import datetime
sba_loan['ApprovalDate'] = pd.to_datetime(sba_loan['ApprovalDate'])
sba_loan['ChgOffDate'] = pd.to_datetime(sba_loan['ChgOffDate'])
sba_loan

LoanNr_ChkDgt                            Name             City State  \
0          1000014003                  ABC HOBBYCRAFT       EVANSVILLE    IN   
1          1000024006     LANDMARK BAR & GRILLE (THE)        NEW PARIS    IN   
2          1000034009           WHITLOCK DDS, TODD M.      BLOOMINGTON    IN   
3          1000044001   BIG BUCKS PAWN & JEWELRY, LLC     BROKEN ARROW    OK   
4          1000054004     ANASTASIA CONFECTIONS, INC.          ORLANDO    FL   
...               ...                             ...              ...   ...   
899159     9995573004                    FABRIC FARMS  UPPER ARLINGTON    OH   
899160     9995603000                    FABRIC FARMS         COLUMBUS    OH   
899161     9995613003    RADCO MANUFACTURING CO.,INC.      SANTA MARIA    CA   
899162     9995973006           MARUTAMA HAWAII, INC.         HONOLULU    HI   
899163     9996003010  PACIFIC TRADEWINDS FAN & LIGHT           KAILUA    HI   

          Zip                            Bank BankState   NAICS ApprovalDate  \
0       47711                FIFTH THIRD BANK        OH  451120   1997-02-28   
1       46526                 1ST SOURCE BANK        IN  722410   1997-02-28   
2       47401         GRANT COUNTY STATE BANK        IN  621210   1997-02-28   
3       74012   1ST NATL BK & TR CO OF BROKEN        OK       0   1997-02-28   
4       32801         FLORIDA BUS. DEVEL CORP        FL       0   1997-02-28   
...       ...                             ...       ...     ...          ...   
899159  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451120   1997-02-27   
899160  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451130   1997-02-27   
899161  93455  RABOBANK, NATIONAL ASSOCIATION        CA  332321   1997-02-27   
899162  96830                  BANK OF HAWAII        HI       0   1997-02-27   
899163  96734            CENTRAL PACIFIC BANK        HI       0   1997-02-27   

       ApprovalFY  ...  ChgOffDate  DisbursementDate  DisbursementGross  \
0            1997  ...         NaT         28-Feb-99            60000.0   
1            1997  ...         NaT         31-May-97            40000.0   
2            1997  ...         NaT         31-Dec-97           287000.0   
3            1997  ...         NaT         30-Jun-97            35000.0   
4            1997  ...         NaT         14-May-97           229000.0   
...           ...  ...         ...               ...                ...   
899159       1997  ...         NaT         30-Sep-97            70000.0   
899160       1997  ...         NaT         31-Oct-97            85000.0   
899161       1997  ...         NaT         30-Sep-97           300000.0   
899162       1997  ...  2000-03-08         31-Mar-97            75000.0   
899163       1997  ...         NaT         31-May-97            30000.0   

        BalanceGross  MIS_Status  ChgOffPrinGr    GrAppv  SBA_Appv  NAICS_2  \
0                0.0       P I F           0.0   60000.0   48000.0       45   
1                0.0       P I F           0.0   40000.0   32000.0       72   
2                0.0       P I F           0.0  287000.0  215250.0       62   
3                0.0       P I F           0.0   35000.0   28000.0        0   
4                0.0       P I F           0.0  229000.0  229000.0        0   
...              ...         ...           ...       ...       ...      ...   
899159           0.0       P I F           0.0   70000.0   56000.0       45   
899160           0.0       P I F           0.0   85000.0   42500.0       45   
899161           0.0       P I F           0.0  300000.0  225000.0       33   
899162           0.0      CHGOFF       46383.0   75000.0   60000.0        0   
899163           0.0       P I F           0.0   30000.0   24000.0        0   

                                 Industry  
0                            Retail trade  
1         Accommodation and food services  
2       Health care and social assistance  
3                                    None  
4               

***

## Submit Your Work

You're almost done -- congratulations!

You need to do a few more things:

1. Save your work.  To do this, create a **notebook checkpoint** by using the menu within the notebook to go **File -> Save and Checkpoint**

2. Click **File** -> **Close and Halt** to close this notebook.

3. Click **Logout** on Jupyter to return your terminal back to the command prompt.

4. Follow the assignment instructions to submit this assignment.